### PR TITLE
Using a data probing strategy, efficiently retrieve the top level tag to determine which top level parser to use

### PR DESCRIPTION
**Pull request by DJBen**
- This change makes all the parsing errors in `partwise` documents MUCH more understandable because it will no longer be swallowed.
- This change also increases parsing speed for `timewise` documents because it will no longer attempt to parse it with `partwise` parser first.

### CODE REVIEW

File: Sources/MusicXML/Decoding/Decoding.swift [hunk 1]
```diff
@@ -9,6 +9,7 @@ import Foundation
 import XMLCoder
 
 extension Score {
+    public static let topLevelTagKey = "topLevelTagKey"
```
Comment (bwetherfield) [inline Sources/MusicXML/Decoding/Decoding.swift:12]: Would it make sense to instead store `let traversalDirection: CodingKeys` in `Score`, bypassing the `userInfo` layer? I am also thinking this property might be helpful when we get to encoding the `Score` value by a similar pattern?

**Comment by bwetherfield** [comment]
Thank you so much for contributing this, @DJBen! Looks amazing. Just the one question I had regarding the use of `XMLDecoder.userInfo` versus storing the relevant `Score.CodingKeys` value directly.

File: Sources/MusicXML/Decoding/Decoding.swift (hunk 1 above)
Comment (DJBen) [reply Sources/MusicXML/Decoding/Decoding.swift:12]: That doesn't work - `self` hasn't been initialized yet. 

The only way I pass information to `public init(from decoder: Decoder) throws` is thru the `decoder`.

**Comment by jsbean** [comment]
This looks great to me!

**Pull request merged**